pkg/auth/handler/reset_password_test.go

```diff
@@ -0,0 +1,136 @@
+package handler
+
+import (
+	"testing"
+
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/provider/password"
+	"github.com/skygeario/skygear-server/pkg/auth/response"
+	coreAudit "github.com/skygeario/skygear-server/pkg/core/audit"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authinfo"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authtoken"
+	"github.com/skygeario/skygear-server/pkg/server/audit"
+	"github.com/skygeario/skygear-server/pkg/server/skyerr"
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+func TestResetPasswordHandler(t *testing.T) {
+	Convey("Test ResetPasswordRequestPayload", t, func() {
+		Convey("validate valid payload", func() {
+			payload := ResetPasswordRequestPayload{
+				AuthInfoID: "1",
+				Password:   "123456",
+			}
+			So(payload.Validate(), ShouldBeNil)
+		})
+
+		Convey("validate payload without auth_id", func() {
+			payload := ResetPasswordRequestPayload{
+				Password: "123456",
+			}
+			err := payload.Validate()
+			errResponse := err.(skyerr.Error)
+			So(errResponse.Code(), ShouldEqual, skyerr.InvalidArgument)
+		})
+
+		Convey("validate payload without password", func() {
+			payload := ResetPasswordRequestPayload{
+				AuthInfoID: "1",
+			}
+			err := payload.Validate()
+			errResponse := err.(skyerr.Error)
+			So(errResponse.Code(), ShouldEqual, skyerr.InvalidArgument)
+		})
+	})
+
+	Convey("Test ResetPasswordHandler", t, func() {
+		// fixture
+		authInfoStore := authinfo.NewMockStoreWithAuthInfoMap(
+			map[string]authinfo.AuthInfo{
+				"john.doe.id": authinfo.AuthInfo{
+					ID: "john.doe.id",
+				},
+			},
+		)
+		passwordAuthProvider := password.NewMockProviderWithPrincipalMap(
+			map[string]password.Principal{
+				"john.doe.principal.id": password.Principal{
+					ID:     "john.doe.principal.id",
+					UserID: "john.doe.id",
+					AuthData: map[string]interface{}{
+						"username": "john.doe",
+						"email":    "john.doe@example.com",
+					},
+					HashedPassword: []byte("$2a$10$/jm/S1sY6ldfL6UZljlJdOAdJojsJfkjg/pqK47Q8WmOLE19tGWQi"), // 123456
+				},
+			},
+		)
+		tokenStore := authtoken.NewJWTStore("myApp", "secret", 0)
+		passwordChecker := &audit.PasswordChecker{
+			PwMinLength: 6,
+		}
+
+		h := &ResetPasswordHandler{}
+		h.AuthInfoStore = authInfoStore
+		h.TokenStore = tokenStore
+		h.PasswordChecker = passwordChecker
+		h.PasswordAuthProvider = passwordAuthProvider
+		h.AuditTrail = coreAudit.NewMockTrail(t)
+
+		Convey("reset password with user id", func() {
```

Comment:
The message format should try to follow the BDD format: https://hiptest.com/blog/2016/04/28/getting-started-with-bdd-part-1/